<?xml version="1.0" encoding="UTF-8"?>
<root BTCPP_format="3"
      main_tree_to_execute="decision">
  <BehaviorTree ID="GlobalStatusPublisher">
    <Sequence>
      <pubRobotStatus hp_status="{hp_status}"
                      cruise_status="{cruise_status}"
                      pursuit_status="{pursuit_status}"
                      target_goal="{target_goal}"
                      is_hp_ok="{is_hp_ok}"
                      pursuit_mode1="{pursuit_mode1}"
                      pursuit_mode2="{pursuit_mode2}"
                      cruise_goal="{cruise_goal}"
                      supply_goal="{supply_goal}"
                      goal_pose_override="{goal_pose_override}"
                      navigate_to_pose_status="{navigate_to_pose_status}"
                      current_controller="{current_controller}"
                      is_in_special_area="{is_in_special_area}"
                      goal_poses="{goal_poses}"
                      do_clear="{do_clear}"/>
    </Sequence>
  </BehaviorTree>

  <BehaviorTree ID="add_def_cru_copy">
    <Parallel failure_threshold="3"
              success_threshold="3">
      <WhileDoElse>
        <Status Status="{hp_status}"/>
        <SubTree ID="addblood"
                 __shared_blackboard="true"/>
        <SubTree ID="defense"
                 __shared_blackboard="true"/>
      </WhileDoElse>
      <SubTree ID="cruise"
               __shared_blackboard="true"/>
      <SubTree ID="defense"
               __shared_blackboard="true"/>
    </Parallel>
  </BehaviorTree>

  <BehaviorTree ID="addblood">
    <PipelineSequence name="NavigateToKeepADistance">
      <RateController hz="50.0">
        <Sequence>
          <GoalUpdater input_goal="{supply_goal}"
                       output_goal="{updated_target_goal}">
            <ComputePathToPose goal="{updated_target_goal}"
                               planner_id="GridBased"
                               path="{goal_path}"/>
          </GoalUpdater>
          <TruncatePath distance="0.1"
                        input_path="{goal_path}"
                        output_path="{truncated_goal_path}"/>
        </Sequence>
      </RateController>
      <KeepRunningUntilFailure>
        <FollowPath controller_id="FollowPath"
                    path="{truncated_goal_path}"/>
      </KeepRunningUntilFailure>
    </PipelineSequence>
  </BehaviorTree>

  <BehaviorTree ID="attack_outpost">
    <WhileDoElse>
      <Inverter>
        <Status Status="{enermy_outpost_status}"/>
      </Inverter>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="attack_outpost"
                        goal="0;map;9.0;4.82;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="15000">
            <Wait wait_duration="15"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <AlwaysSuccess/>
    </WhileDoElse>
  </BehaviorTree>

  <BehaviorTree ID="attack_outpost_2">
    <WhileDoElse>
      <Inverter>
        <Status Status="{enermy_outpost_status}"/>
      </Inverter>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="attack_outpost"
                        goal="0;map;6.7;6.9;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="15000">
            <Wait wait_duration="15"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <AlwaysSuccess/>
    </WhileDoElse>
  </BehaviorTree>

  <BehaviorTree ID="bullet_supply">
    <WhileDoElse>
      <Status Status="{bullet_status}"/>
      <Parallel name="add bullet"
                failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="回补血点"
                        goal="0;map;-3.41;-6.23;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="30000">
            <Wait wait_duration="30"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <AlwaysSuccess/>
    </WhileDoElse>
  </BehaviorTree>

  <BehaviorTree ID="cruise">
    <PipelineSequence name="NavigateToKeepADistance">
      <RateController hz="50.0">
        <Sequence>
          <GoalUpdater input_goal="{cruise_goal}"
                       output_goal="{updated_target_goal}">
            <ComputePathToPose goal="{updated_target_goal}"
                               planner_id="GridBased"
                               path="{goal_path}"/>
          </GoalUpdater>
          <TruncatePath distance="0.1"
                        input_path="{goal_path}"
                        output_path="{truncated_goal_path}"/>
        </Sequence>
      </RateController>
      <KeepRunningUntilFailure>
        <FollowPath controller_id="{current_controller}"
                    path="{truncated_goal_path}"/>
      </KeepRunningUntilFailure>
    </PipelineSequence>
  </BehaviorTree>

  <BehaviorTree ID="decision">
    <ReactiveSequence>
      <GameStart/>
      <pubRobotStatus hp_status="{hp_status}"
                      cruise_status="{cruise_status}"
                      pursuit_status="{pursuit_status}"
                      target_goal="{target_goal}"
                      is_hp_ok="{is_hp_ok}"
                      pursuit_mode1="{pursuit_mode1}"
                      pursuit_mode2="{pursuit_mode2}"
                      cruise_goal="{cruise_goal}"
                      supply_goal="{supply_goal}"
                      goal_pose_override="{goal_pose_override}"
                      navigate_to_pose_status="{navigate_to_pose_status}"
                      current_controller="{current_controller}"
                      is_in_special_area="{is_in_special_area}"
                      goal_poses="{goal_poses}"
                      do_clear="{do_clear}"/>
      <Parallel failure_threshold="2"
                success_threshold="2">
        <WhileDoElse>
          <Status Status="{navigate_to_pose_status}"/>
          <SubTree ID="navigate_override_subtree"
                   __shared_blackboard="true"/>
          <WhileDoElse>
            <Status Status="{hp_status}"/>
            <SubTree ID="addblood"
                     __shared_blackboard="true"/>
            <WhileDoElse>
              <Status Status="{pursuit_status}"/>
              <SubTree ID="pursuit"
                       __shared_blackboard="true"/>
              <WhileDoElse>
                <Status Status="{cruise_status}"/>
                <SubTree ID="cruise"
                         __shared_blackboard="true"/>
                <AlwaysSuccess/>
              </WhileDoElse>
            </WhileDoElse>
          </WhileDoElse>
        </WhileDoElse>
        <WhileDoElse>
          <Status Status="{do_clear}"/>
          <ClearEntireCostmap service_name="local_costmap/clear_entirely_local_costmap"
                              server_timeout="1000"/>
          <AlwaysSuccess/>
        </WhileDoElse>
      </Parallel>
    </ReactiveSequence>
  </BehaviorTree>

  <BehaviorTree ID="defense">
    <WhileDoElse>
      <Status Status="{defense_status}"/>
      <WhileDoElse>
        <Status Status="{defense_mode0}"/>
        <Sequence>
          <Parallel failure_threshold="2"
                    success_threshold="1">
            <NavigateToPose goal="0;map;4.83;2.40;0;0;0;0;1"/>
            <Inverter>
              <Timeout msec="10000">
                <Wait wait_duration="10"/>
              </Timeout>
            </Inverter>
          </Parallel>
          <Parallel failure_threshold="2"
                    success_threshold="1">
            <NavigateToPose name="雷达导航目标点"
                            goal="0;map;4.363;-0.40;0;0;0;0;1"/>
            <Inverter>
              <Timeout msec="10000">
                <Wait wait_duration="10"/>
              </Timeout>
            </Inverter>
          </Parallel>
        </Sequence>
        <WhileDoElse>
          <Status Status="{defense_mode1}"/>
          <WhileDoElse>
            <Status Status="{pursuit_status}"/>
            <SubTree ID="pursuit"
                     __shared_blackboard="true"/>
            <SubTree ID="基地-启动区巡航defense"
                     name="基地-环高巡航defense"
                     __shared_blackboard="true"/>
          </WhileDoElse>
          <SubTree ID="巡航"
                   __shared_blackboard="true"/>
        </WhileDoElse>
      </WhileDoElse>
      <AlwaysSuccess/>
    </WhileDoElse>
  </BehaviorTree>

  <BehaviorTree ID="defense_outpost">
    <WhileDoElse>
      <Status Status="{cruise_status}"/>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="己方前哨站增益点&#10;"
                        goal="0;map;7.06;-4.928;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="15000">
            <Wait wait_duration="15"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <AlwaysSuccess/>
    </WhileDoElse>
  </BehaviorTree>

  <BehaviorTree ID="navigate_override_subtree">
    <PipelineSequence name="NavigateToPoseOveride">
      <RateController hz="50.0">
        <Sequence>
          <GoalUpdater input_goal="{goal_pose_override}"
                       output_goal="{updated_target_goal_override}">
            <ComputePathToPose goal="{updated_target_goal_override}"
                               planner_id="GridBased"
                               path="{goal_path_override}"/>
          </GoalUpdater>
          <TruncatePath distance="0.01"
                        input_path="{goal_path_override}"
                        output_path="{truncated_goal_path_override}"/>
        </Sequence>
      </RateController>
      <KeepRunningUntilFailure>
        <FollowPath controller_id="{current_controller}"
                    path="{truncated_goal_path_override}"/>
      </KeepRunningUntilFailure>
    </PipelineSequence>
  </BehaviorTree>

  <BehaviorTree ID="pursuit">
    <PipelineSequence name="NavigateWithReplanning">
      <RateController hz="50.0">
        <Sequence>
          <GoalUpdater input_goal="{target_goal}"
                       output_goal="{updated_target_goal}">
            <ComputePathToPose goal="{updated_target_goal}"
                               planner_id="GridBased"
                               path="{goal_path}"/>
          </GoalUpdater>
          <TruncatePath distance="2.6"
                        input_path="{goal_path}"
                        output_path="{truncated_goal_path}"/>
        </Sequence>
      </RateController>
      <KeepRunningUntilFailure>
        <FollowPath controller_id="FollowPath"
                    path="{truncated_goal_path}"/>
      </KeepRunningUntilFailure>
    </PipelineSequence>
  </BehaviorTree>

  <BehaviorTree ID="上R3test">
    <Sequence>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方R3高地下方中继点"
                        goal="0;map;11.01;-2.951;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方R3高地上中继点"
                        goal="0;map;10.13;-8.12;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方英雄狙击点"
                        goal="0;map;7.3;-7.19;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方R3高地上中继点"
                        goal="0;map;10.13;-8.12;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方R3高地下方中继点"
                        goal="0;map;11.01;-2.951;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
    </Sequence>
  </BehaviorTree>

  <BehaviorTree ID="上敌方环高">
    <Sequence>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="  敌方环高中继点（靠近R3高地）&#10;"
                        goal="0;map;12.56;-7.22;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name=" 敌方环高中继点（靠近飞坡禁区）"
                        goal="0;map;9.29;-6.95;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方环高增益点"
                        goal="0;map;12.23;-2.79;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name=" 敌方环高中继点（靠近飞坡禁区）"
                        goal="0;map;9.29;-6.95;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="  敌方环高中继点（靠近R3高地）&#10;"
                        goal="0;map;12.56;-7.22;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
    </Sequence>
  </BehaviorTree>

  <BehaviorTree ID="上英雄吊射点">
    <Sequence>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方R3高地下方中继点"
                        goal="0;map;18.57;-1.51;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方R3高地上中继点"
                        goal="0;map;18.21;-5.72;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方英雄狙击点"
                        goal="0;map;16.67;-6.78;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方R3高地上中继点"
                        goal="0;map;18.21;-5.72;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="     敌方R3高地下方中继点"
                        goal="0;map;18.57;-1.51;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="5000">
            <Wait wait_duration="5"/>
          </Timeout>
        </Inverter>
      </Parallel>
    </Sequence>
  </BehaviorTree>

  <BehaviorTree ID="决策1-过荒地">
    <ReactiveSequence>
      <SubTree ID="attack_outpost"
               __shared_blackboard="true"/>
      <Status Status="{enermy_outpost_status}"/>
      <Parallel failure_threshold="4"
                success_threshold="4">
        <SubTree ID="addblood"
                 __shared_blackboard="true"/>
        <SubTree ID="cruise"
                 __shared_blackboard="true"/>
        <SubTree ID="defense"
                 __shared_blackboard="true"/>
        <SubTree ID="pursuit"
                 __shared_blackboard="true"/>
      </Parallel>
    </ReactiveSequence>
  </BehaviorTree>

  <BehaviorTree ID="决策2-英雄前面击打前哨">
    <ReactiveSequence>
      <SubTree ID="attack_outpost_2"
               __shared_blackboard="true"/>
      <Status Status="{enermy_outpost_status}"/>
      <Parallel failure_threshold="4"
                success_threshold="4">
        <SubTree ID="addblood"
                 __shared_blackboard="true"/>
        <SubTree ID="defense_outpost"
                 __shared_blackboard="true"/>
        <SubTree ID="defense"
                 __shared_blackboard="true"/>
        <SubTree ID="pursuit"
                 __shared_blackboard="true"/>
      </Parallel>
    </ReactiveSequence>
  </BehaviorTree>

  <BehaviorTree ID="基地-启动区巡航defense">
    <Sequence>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="基地左边"
                        goal="0;map;-3.02;1.8923;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="10000">
            <Wait wait_duration="10"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="基地左边"
                        goal="0;map;-3.02;-1.923;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="10000">
            <Wait wait_duration="10"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="base defense point"
                        goal="0;map;0.5;1.50;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="10000">
            <Wait wait_duration="10"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="base defense point"
                        goal="0;map;-0.5;-1.50;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="10000">
            <Wait wait_duration="10"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="base defense point"
                        goal="0;map;0.5;-1.50;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="10000">
            <Wait wait_duration="10"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="base defense point"
                        goal="0;map;-0.5;1.50;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="10000">
            <Wait wait_duration="10"/>
          </Timeout>
        </Inverter>
      </Parallel>
    </Sequence>
  </BehaviorTree>

  <BehaviorTree ID="巡航">
    <PipelineSequence>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="base defense point"
                        goal="0;map;4.0;-4.0;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="10000">
            <Wait wait_duration="10"/>
          </Timeout>
        </Inverter>
      </Parallel>
      <Parallel failure_threshold="2"
                success_threshold="1">
        <NavigateToPose name="base defense point"
                        goal="0;map;0.158;-0.0;0;0;0;0;1"/>
        <Inverter>
          <Timeout msec="10000">
            <Wait wait_duration="10"/>
          </Timeout>
        </Inverter>
      </Parallel>
    </PipelineSequence>
  </BehaviorTree>

  <BehaviorTree ID="敌方哨兵启动点（作为地方半场的中继点）">
    <Parallel failure_threshold="2"
              success_threshold="1">
      <NavigateToPose name="     敌方R3高地下方中继点"
                      goal="0;map;17.83;0.0;0;0;0;0;1"/>
      <Inverter>
        <Timeout msec="10000">
          <Wait wait_duration="8"/>
        </Timeout>
      </Inverter>
    </Parallel>
  </BehaviorTree>

  <!-- Description of Node Models (used by Groot) -->
  <TreeNodesModel>
    <Action ID="ClearEntireCostmap">
      <input_port name="service_name">Service name</input_port>
      <input_port name="server_timeout">Server timeout</input_port>
    </Action>
    <Action ID="ComputePathToPose"
            editable="true">
      <input_port name="goal">Destination to plan to</input_port>
      <input_port name="planner_id">Mapped name to the planner plugin type to use</input_port>
      <output_port name="path">Path created by ComputePathToPose node</output_port>
    </Action>
    <Action ID="FollowPath"
            editable="true">
      <input_port name="controller_id"
                  default="FollowPath"/>
      <input_port name="path">Path to follow</input_port>
    </Action>
    <Condition ID="GameStart"
               editable="true"/>
    <Decorator ID="GoalUpdater">
      <input_port name="input_goal">Original goal in</input_port>
      <output_port name="output_goal">Output goal set by subscription</output_port>
    </Decorator>
    <Action ID="NavigateToPose">
      <input_port name="goal">Goal</input_port>
    </Action>
    <Control ID="PipelineSequence"/>
    <Decorator ID="RateController">
      <input_port name="hz">Rate</input_port>
    </Decorator>
    <Condition ID="Status"
               editable="true">
      <input_port name="Status"
                  default="{Status}"/>
    </Condition>
    <Action ID="TruncatePath">
      <input_port name="distance">Distance before goal to truncate</input_port>
      <input_port name="input_path">Path to truncate</input_port>
      <output_port name="output_path">Truncated path to utilize</output_port>
    </Action>
    <Action ID="Wait">
      <input_port name="wait_duration">Wait time</input_port>
    </Action>
    <Action ID="pubRobotStatus"
            editable="true">
      <output_port name="hp_status"
                   default="{hp_status}"/>
      <output_port name="cruise_status"
                   default="{cruise_status}"/>
      <output_port name="pursuit_status"
                   default="{pursuit_status}"/>
      <output_port name="target_goal"
                   default="{target_goal}"/>
      <output_port name="is_hp_ok"
                   default="{is_hp_ok}"/>
      <output_port name="pursuit_mode1"
                   default="{pursuit_mode1}"/>
      <output_port name="pursuit_mode2"
                   default="{pursuit_mode2}"/>
      <output_port name="cruise_goal"
                   default="{cruise_goal}"/>
      <output_port name="supply_goal"
                   default="{supply_goal}"/>
      <output_port name="goal_pose_override"
                   default="{goal_pose_override}"/>
      <output_port name="navigate_to_pose_status"
                   default="{navigate_to_pose_status}"/>
      <output_port name="current_controller"
                   default="{current_controller}"/>
      <output_port name="is_in_special_area"
                   default="{is_in_special_area}"/>
      <output_port name="goal_poses"
                   default="{goal_poses}"/>
      <output_port name="do_clear"
                   default="{do_clear}"/>
    </Action>
  </TreeNodesModel>

</root>
